### PR TITLE
Add ability to remove promotions via Promotion#remove_from

### PR DESCRIPTION
**Pull request by jordan-brough**
Extracted from #1433 as per a chat with @jhawthorn.

This will allow promotions to be removed from orders and allows promotion actions to define how to reverse their side effects on an order.

For now this provides a default `remove_from` on the `PromotionAction` base class, with a deprecation warning that subclasses should define their own `remove_from` method.

### CODE REVIEW

File: core/app/models/spree/promotion.rb [hunk 1]
```diff
@@ -220,6 +220,20 @@ def used_by?(user, excluded_orders = [])
       end
     end
 
+    # Removes a promotion and any adjustments or other side effects from an
+    # order.
+    # @param order [Spree::Order] the order to remove the promotion from.
+    # @return [undefined]
```
Comment (gmacdougall) [inline core/app/models/spree/promotion.rb]: If you don't care about the return value `@return [void]` should be used.

File: core/app/models/spree/promotion/actions/create_adjustment.rb [hunk 2]
```diff
@@ -39,6 +39,16 @@ def compute_amount(calculable)
           [(calculable.item_total + calculable.ship_total), amount].min * -1
         end
 
+        # Removes any adjustments generated by this action from the order.
+        # @param order [Spree::Order] the order to remove the action from.
```
Comment (gmacdougall) [inline core/app/models/spree/promotion/actions/create_adjustment.rb]: Same here.

File: core/spec/models/spree/promotion/actions/create_item_adjustments_spec.rb [hunk 3]
```diff
@@ -4,11 +4,11 @@ module Spree
   class Promotion
     module Actions
       describe CreateItemAdjustments, type: :model do
-        let(:order) { create(:order) }
+        let(:order) { create(:order_with_line_items, line_items_count: 1) }
         let(:promotion) { create(:promotion, :with_line_item_adjustment, adjustment_rate: adjustment_amount) }
         let(:adjustment_amount) { 10 }
         let(:action) { promotion.actions.first! }
-        let!(:line_item) { create(:line_item, order: order) }
+        let(:line_item) { order.line_items.to_a.first }
```
Comment (gmacdougall) [inline core/spec/models/spree/promotion/actions/create_item_adjustments_spec.rb]: The `to_a` shouldn't be needed

File: core/spec/models/spree/promotion/actions/create_adjustment_spec.rb [hunk 4]
```diff
@@ -58,6 +58,27 @@
     end
   end
 
+  describe '#remove_from' do
+    let(:action) { promotion.actions.first! }
+    let(:promotion) { create(:promotion, :with_order_adjustment) }
+
+    # this adjustment should not get removed
```
Comment (gmacdougall) [inline core/spec/models/spree/promotion/actions/create_adjustment_spec.rb]: Could we name the `other_adjustment` something that implies this more directly?

File: core/app/models/spree/promotion.rb (hunk 1 above)
Comment (jordan-brough) [inline core/app/models/spree/promotion.rb]: @gmacdougall thanks! Updated and opened a PR to remove more of these: https://github.com/solidusio/solidus/pull/1454

File: core/app/models/spree/promotion/actions/create_adjustment.rb (hunk 2 above)
Comment (jordan-brough) [inline core/app/models/spree/promotion/actions/create_adjustment.rb]: Updated. Thanks.

File: core/spec/models/spree/promotion/actions/create_item_adjustments_spec.rb (hunk 3 above)
Comment (jordan-brough) [inline core/spec/models/spree/promotion/actions/create_item_adjustments_spec.rb]: @gmacdougall this ensures that `line_item` refers to the same object that `OrderUpdater` and the promo code & etc will modify.  Our current implementation of the order factory happens to preload `order.line_items` properly, but that's not tested and I'm not sure it was intentional either.  Perhaps I could remove it here but make an additional PR that tests this explicitly in `order_factory_spec`?

File: core/spec/models/spree/promotion/actions/create_adjustment_spec.rb (hunk 4 above)
Comment (jordan-brough) [inline core/spec/models/spree/promotion/actions/create_adjustment_spec.rb]: @gmacdougall sure, I've updated it to `unrelated_adjustment`.

File: core/spec/models/spree/promotion/actions/create_item_adjustments_spec.rb (hunk 3 above)
Comment (jhawthorn) [inline core/spec/models/spree/promotion/actions/create_item_adjustments_spec.rb]: I'm also prefer keeping the `to_a.first`

Comment (gmacdougall) [inline core/spec/models/spree/promotion/actions/create_item_adjustments_spec.rb]: Sounds good

**Comment by flyfy1** [comment]
Currently I can see the `remove_from` being defined.. however, where is this method called? I searched through the solidus core but I cannot find a single place where this `remove_from` method being called.

**Pull request merged**